File: admin-frontend/src/services/sentryConfig.ts
/**
 * Sentry 配置管理服务
 */

import axios from '@/utils/axios';

export interface SentryConfig {
  id: number;
  dsn: string;
  environment: string;
  frontend_enabled: boolean;
  admin_frontend_enabled: boolean;
  traces_sample_rate: string;
  replays_session_sample_rate: string;
  replays_on_error_sample_rate: string;
  ignore_errors?: string;
  allowed_urls?: string;
  denied_urls?: string;
  release_version?: string;
  debug_mode: boolean;
  attach_stacktrace: boolean;
  description?: string;
  created_at: string;
  updated_at: string;
  created_by?: number;
  updated_by?: number;
}

export interface SentryConfigCreate {
  dsn: string;
  environment?: string;
  frontend_enabled?: boolean;
  admin_frontend_enabled?: boolean;
  traces_sample_rate?: string;
  replays_session_sample_rate?: string;
  replays_on_error_sample_rate?: string;
  ignore_errors?: string;
  allowed_urls?: string;
  denied_urls?: string;
  release_version?: string;
  debug_mode?: boolean;
  attach_stacktrace?: boolean;
  description?: string;
}

export interface SentryConfigUpdate {
  dsn?: string;
  environment?: string;
  frontend_enabled?: boolean;
  admin_frontend_enabled?: boolean;
  traces_sample_rate?: string;
  replays_session_sample_rate?: string;
  replays_on_error_sample_rate?: string;
  ignore_errors?: string;
  allowed_urls?: string;
  denied_urls?: string;
  release_version?: string;
  debug_mode?: boolean;
  attach_stacktrace?: boolean;
  description?: string;
}

// 获取所有配置
export const getSentryConfigs = async (): Promise<SentryConfig[]> => {
  const response = await axios.get('/api/v1/admin/sentry-config/');
  return response.data;
};

// 获取单个配置
export const getSentryConfig = async (id: number): Promise<SentryConfig> => {
  const response = await axios.get(`/api/v1/admin/sentry-config/${id}`);
  return response.data;
};

// 获取当前激活的配置
export const getActiveSentryConfig = async (): Promise<SentryConfig> => {
  const response = await axios.get('/api/v1/admin/sentry-config/active/current');
  return response.data;
};

// 创建配置
export const createSentryConfig = async (data: SentryConfigCreate): Promise<SentryConfig> => {
  const response = await axios.post('/api/v1/admin/sentry-config/', data);
  return response.data;
};

// 更新配置
export const updateSentryConfig = async (
  id: number,
  data: SentryConfigUpdate
): Promise<SentryConfig> => {
  const response = await axios.put(`/api/v1/admin/sentry-config/${id}`, data);
  return response.data;
};

// 删除配置
export const deleteSentryConfig = async (id: number): Promise<void> => {
  await axios.delete(`/api/v1/admin/sentry-config/${id}`);
};

File: backend/app/schemas/sentry_config.py
"""
Sentry 配置相关的 Pydantic schemas
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, validator


class SentryConfigBase(BaseModel):
    """Sentry 配置基础模型"""

    dsn: str = Field(..., description="Sentry DSN")
    environment: str = Field(default="production", description="环境名称")
    frontend_enabled: bool = Field(default=True, description="用户前端是否启用")
    admin_frontend_enabled: bool = Field(default=True, description="管理前端是否启用")
    traces_sample_rate: str = Field(default="1.0", description="性能监控采样率")
    replays_session_sample_rate: str = Field(default="0.1", description="会话回放采样率")
    replays_on_error_sample_rate: str = Field(default="1.0", description="错误回放采样率")
    ignore_errors: Optional[str] = Field(None, description="忽略的错误列表（JSON数组）")
    allowed_urls: Optional[str] = Field(None, description="允许上报的URL列表（JSON数组）")
    denied_urls: Optional[str] = Field(None, description="拒绝上报的URL列表（JSON数组）")
    release_version: Optional[str] = Field(None, description="发布版本号")
    debug_mode: bool = Field(default=False, description="调试模式")
    attach_stacktrace: bool = Field(default=True, description="自动附加堆栈跟踪")
    description: Optional[str] = Field(None, description="配置说明")

    @validator("traces_sample_rate", "replays_session_sample_rate", "replays_on_error_sample_rate")
    def validate_sample_rate(cls, v):
        """验证采样率在 0.0-1.0 之间"""
        try:
            rate = float(v)
            if not 0.0 <= rate <= 1.0:
                raise ValueError("采样率必须在 0.0 到 1.0 之间")
            return str(rate)
        except ValueError:
            raise ValueError("采样率必须是有效的数字")


class SentryConfigCreate(SentryConfigBase):
    """创建 Sentry 配置"""

    pass


class SentryConfigUpdate(BaseModel):
    """更新 Sentry 配置"""

    dsn: Optional[str] = None
    environment: Optional[str] = None
    frontend_enabled: Optional[bool] = None
    admin_frontend_enabled: Optional[bool] = None
    traces_sample_rate: Optional[str] = None
    replays_session_sample_rate: Optional[str] = None
    replays_on_error_sample_rate: Optional[str] = None
    ignore_errors: Optional[str] = None
    allowed_urls: Optional[str] = None
    denied_urls: Optional[str] = None
    release_version: Optional[str] = None
    debug_mode: Optional[bool] = None
    attach_stacktrace: Optional[bool] = None
    description: Optional[str] = None

    @validator("traces_sample_rate", "replays_session_sample_rate", "replays_on_error_sample_rate")
    def validate_sample_rate(cls, v):
        """验证采样率在 0.0-1.0 之间"""
        if v is None:
            return v
        try:
            rate = float(v)
            if not 0.0 <= rate <= 1.0:
                raise ValueError("采样率必须在 0.0 到 1.0 之间")
            return str(rate)
        except ValueError:
            raise ValueError("采样率必须是有效的数字")


class SentryConfigResponse(SentryConfigBase):
    """Sentry 配置响应"""

    id: int
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    updated_by: Optional[int] = None

    class Config:
        from_attributes = True


class SentryConfigPublic(BaseModel):
    """Sentry 公开配置（用于前端获取）"""

    dsn: str
    environment: str
    traces_sample_rate: str
    replays_session_sample_rate: str
    replays_on_error_sample_rate: str
    release_version: Optional[str] = None
    debug_mode: bool
    attach_stacktrace: bool
    ignore_errors: Optional[str] = None
    allowed_urls: Optional[str] = None
    denied_urls: Optional[str] = None

    class Config:
        from_attributes = True
